Implement broadcasting a pre-image

linked0 opened the pull request:
This PR is about broadcasting a pre-image and saving it into the validator set table. 

- sendPreimage endpoint in Validator API

#492

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -35,6 +35,19 @@ import std.path;
+{
+    public Hash utxo_key;
+    public Hash hash;
+    public uint round;
```
That's an interesting approach. When we were originally talking about it, I had in mind that this would by *block height*, not round. Because the issue is that if you re-use the same `utxo_key` (as you would), then you can do replay attacks.
Of course validation would fail, but you could potentially pretend a node sent you the wrong preimage data. We didn't factor any slashing rules that could be triggered by this tho.

A downside of using the `block_height` is that it's harder to pre-emptively broadcast preimage as you don't know whe height at which it'll be included. But that can be dealt with when the enrollment is received.

Geod24 replied:
In any case, this could use some documentation

linked0 replied:
I'd like accept your thought because it's more convenient to recognize the block to which the preimage is related. So, I could add the variable named `start_block_height` into the Enrollment structure which is the block height where a validator is supposed to participate in validating. Then, The variable, `round` of the PreimageInfo structure could be replaced with a variable like `round_block_height`. How about this?

Geod24 replied:
> start_block_height

Minor stylistic note: "height" is pretty unambiguous in blockchain code, so you can just use `start_height` (because there's no other height that block height).

> which is the block height where a validator is supposed to participate in validating

I'll have to double check, the exact flow, but bear in mind that if we go this way, it is an *implicit* value. What I mean is that, when you create an `Enrollment` structure and sign it, you cannot know which `start_height` you will be assigned, because it has to be accepted by the validators. Same goes for transactions, when you create a transaction, you cannot know at which height it will be.

Does that make sense ? Also, I think both approaches have their merits, so maybe let's have a whiteboarding session about it ?

linked0 replied:
I understand what you're considering. After a bit of thought, I can use the `enrolled_height` variable which is implemented by Henry. The value of `enrolled_height + 1` will have the same value as what I want to have with the `start_block_height` variable. It's good to have a whiteboarding session.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -670,4 +708,10 @@ unittest
+    PreimageInfo preimage;
+    auto round = 1000;
+    assert(man.getPreimage(1000, preimage));
+    assert(preimage.hash == man.preimages[round-1]);
```
Should also test with `0` and > `ValidatorCycleLength`

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -1400,3 +1446,26 @@ unittest
+///
+unittest
+{
+    uint CycleLength = 1008;
```
Use `const` if it isn't going to be modified

Geod24 replied:
Also, since you made it a constant in an earlier commit, can we just use it ?

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -948,6 +948,44 @@ public string isInvalidEnrollmentReason (const ref Enrollment enrollment,
+*******************************************************************************/
+
+public string isInvalidPreimageReason (const ref Hash preimage, const uint round,
+    const ref Hash init_seed, immutable uint cycle_length) nothrow @safe
```
For value type, `const` / `immutable` does not make sense. So just use `uint`.

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -948,6 +948,44 @@ public string isInvalidEnrollmentReason (const ref Enrollment enrollment,
+
+    // check the pre-image has a right hash value
+    Hash temp_hash = preimage;
+    for (int i = round; i < cycle_length; i++)
```
```suggestion
    for (uint i = round; i < cycle_length; i++)
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -522,18 +523,164 @@ public class EnrollmentManager
             return false;
 
-        preimage.utxo_key = this.data.utxo_key;
+        preimage.enroll_key = this.data.utxo_key;
```
Those fixes should be in a previous commit

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -522,18 +523,164 @@ public class EnrollmentManager
+    }
+
+    /***************************************************************************
+        Add a pre-image information to a validator data
```
```suggestion

        Add a pre-image information to a validator data
```

Geod24 commented on source/agora/network/NetworkClient.d:
```diff
@@ -297,4 +298,24 @@ class NetworkClient
+
+    ***************************************************************************/
+
+    public void sendPreimage (PreimageInfo preimage) @trusted
```
```suggestion
    public void sendPreimage (ref PreimageInfo preimage) @trusted
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -35,6 +35,19 @@ import std.path;
+
+*******************************************************************************/
+
+public struct PreimageInfo
```
This should probably be in `consensus.data`

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -433,6 +433,12 @@ public interface TestAPI : API
     public abstract Enrollment createEnrollmentData();
+
+    ///
+    public abstract PreimageInfo getNodePreimage(uint round);
```
```suggestion
    public abstract PreimageInfo getNodePreimage (uint round);
```

Geod24 commented on source/agora/api/Validator.d:
```diff
@@ -93,4 +94,16 @@ public interface API : agora.api.FullNode.API
+
+    ***************************************************************************/
+
+    public void sendPreimage (PreimageInfo preimage);
```
Naming wise, we have `receiveEnvelope` but `sendPreimage`. I don't mind either way but we should be consistent.

linked0 replied:
Do you mean changing the function name into a name like receivePreimage? If you do, changing the name is not a problem for me. I mean I can change the neme.

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -75,4 +75,12 @@ unittest
+    preimage_1 = node_1.getNodePreimage(1000);
+    node_1.sendPreimage(preimage_1);
+    auto preimage_2 = node_2.getValidatorPreimage(preimage_1.enroll_key);
+    assert(preimage_1 == preimage_2);
```
This is not black box testing. Adding 2 special methods means you aren't really testing the API.
A better test would be to send a pre-image and check if the (X) next block contains it.
Additionally you could filter the `sendPreimage` function of a block to see if it propagates correctly even in the absence of the preimage in one block (but that depends on Henry's PR)/

linked0 replied:
My original approach is not to save pre-images into a block. It's stored only in the validator set table. A pre-image can be restored from another nodes if the database has been crashed. In that case, another API endpoint like getPreimage is needed with the getNodePreimage and getValidatorPreimage removed. Anyway the two functions could be confusing.

linked0 replied:
There is another approach about saving pre-images. It's to store the source value of the random seed encrypted with a node's private key into the Enrollment structure. With that, if the database crashed, the source value of the random seed used for enrollment can be restored from the chain.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -496,6 +497,31 @@ public class EnrollmentManager
+    ***************************************************************************/
+
+    public bool getPreimage (uint round, out PreimageInfo preimage)
+        @trusted nothrow
```
Why `@trusted` ?

linked0 commented:
I will make a smaller PR including two commits of this PR and apply reviewer's suggestion in the new PR. And then, I will handle with the remaining commits of this PR which is not inluded in the new PR.

Geod24 commented on source/agora/api/Validator.d:
```diff
@@ -43,6 +43,8 @@
 module agora.api.Validator;
 
 import agora.common.crypto.Key;
+import agora.common.Hash;
```
```suggestion
import agora.common.Types;
```

linked0 commented:
Ready for review.
In order to make smaller PRs, I am splitting this PR to three or four PRs inclusive of this PR. So In this PR, I only put the code about the sendPreimage endpoint in Validator API.

TrustHenry commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -755,3 +870,47 @@ unittest
+    assert(man.addPreimage(preimage));
+    assert(man.hasPreimage(utxo_hash, 1));
+    assert(!man.hasPreimage(utxo_hash, 10));
+}
```
`}
`
Style guide last line is required.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
```
The `toString()` call is allocating. I think something like this could work:

```D
static char[] cmd;
cmd.length = 0;
assumeSafeAppend(cmd);
sformat(cmd, "SELECT preimage from validator_set WHERE key = %s", enroll_key);
auto results = this.db.execute(cmd);
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
+        if (!results.empty && results.oneValue!(byte[]).length != 0)
```
If the result is not empty but the array length is zero, doesn't that mean there is a bug?

I think this should be turned into an assert.

```D
auto value = results.oneValue!(byte[]);
assert(value.length != 0);
auto loaded_image = value.deseralizeFull!PreimageInfo;
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+
+    public bool addPreimage (const ref PreimageInfo preimage) @safe nothrow
+    {
+        static ubyte[] buffer;
```
This needs to be moved right before the `buffer.length = 0;` statement. Declarations should be placed close to where they're used the first time, which is not necessarily at the start of the function.

AndrejMitrovic reviewed (changes requested):
Not done reviewing, github won't let me continue because of a push.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+        }
+        catch (Exception ex)
+        {
+            log.error("Serializtion error, addPreimage:{}, exception:{}",
```
typo: serializtion

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+        {
+            () @trusted {
+                this.db.execute("UPDATE validator_set SET preimage = ? " ~
+                    "WHERE key = ?", buffer, preimage.enroll_key.toString());
```
Same comment about memory allocation, `toString()` allocates. Can use `sformat` to write the entire string to a buffer before calling `execute`.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
```
Ok, so when calling `addPreimage` the function is `nothrow` and can fail so it returns a status code. But if `hasPreimage` fails it will throw an exception. This is an inconsistent API. `hasPreimage` should be `nothrow` too.

AndrejMitrovic reviewed (changes requested):
Reviewed first commit.

AndrejMitrovic commented on source/agora/network/NetworkClient.d:
```diff
@@ -297,4 +298,24 @@ class NetworkClient
+    /***************************************************************************
+
+        Send a preimage asynchronously to the node.
+        Any errors are reporteed to the debugging log.
```
```suggestion
        Any errors are reported to the debugging log.
```

AndrejMitrovic commented on source/agora/node/GossipProtocol.d:
```diff
@@ -117,4 +118,22 @@ public class GossipProtocol
+
+    /***************************************************************************
+
+        If there is not the same pre-image on the enrollment manager,
```
```suggestion
        If this is a new enrollment, add it to the enrollment manager and propagate it to the network.
```

AndrejMitrovic commented on source/agora/node/GossipProtocol.d:
```diff
@@ -117,4 +118,22 @@ public class GossipProtocol
+        propagate it to the network.
+
+        Params:
+            enroll = the pre-image information to propagate
```
```suggestion
            enroll = the pre-image information to add & propagate
```

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -515,6 +519,14 @@ public class TestNode : Node, TestAPI
     }
+
+    /// Get a node's own pre-image information used when revealing it
+    public override PreimageInfo getPreimage(uint height)
```
```suggestion
    public override PreimageInfo getPreimage (uint height)
```

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -74,4 +75,12 @@ unittest
+
+    // tests for revealing a pre-image
+    auto preimage = node_1.getPreimage(1000);
+    node_1.receivePreimage(preimage);
```
This node should already know about its own preimages. It is odd that it would need to send the preimage to itself in order to know about it, no?

AndrejMitrovic commented:
Done with review.

linked0 committed:
Add addPreimage and hasPreimage functions to EnrollmentManager

The EnrollmentManger class must provide functions to add a piece of
pre-image information and check existence of it for broadcasting preimage.

linked0 committed:
Add receivePreimage function into the Validator API

A Node reveals a pre-image information into the network. The broadcasting of
a pre-image is processed through the gossip protocol.

linked0 committed:
Add unittest code for broadcasting a preimage through the network

Geod24 reviewed (approved):
LGTM, let's move forward with this.
